### ISSUE TITLE
device of noise models

### ISSUE THREAD

**Issue by tachella** [deepinv/deepinv#634]
There are some inconsistencies when setting the device of the parameters of a noise model. The issue is that the device is decided by the random seed `torch.Generator` input of the noise model, instead of the device of the input parameter.

```
import deepinv as dinv
import torch

physics = dinv.physics.Denoising(noise_model=dinv.physics.GaussianNoise(sigma=0.01), device='cuda')
print(physics.noise_model.sigma.device)
# cpu while I would expect cuda

physics = dinv.physics.Denoising(noise_model=dinv.physics.GaussianNoise(sigma=torch.tensor(0.01, device='cuda')), device='cuda')
print(physics.noise_model.sigma.device)

# cpu while I would expect cuda

generator = torch.Generator(device='cuda')
physics = dinv.physics.Denoising(noise_model=dinv.physics.GaussianNoise(sigma=0.01, rng=generator), device='cuda')
print(physics.noise_model.sigma.device)
# cuda - OK
```

I believe that most users don't use the random seed, and thus can have unexpected device errors that are hard to spot. I propose that the device is decided by the input parameter instead of the generator.

**Comment by Andrewwango** [comment]
Note we can do `physics.to(device)` which is a top-down way of solving this problem.

**Comment by jscanvic** [comment]
Proper type hinting would probably have helped here. It's not obvious in the code that sigma can be a torch.Tensor and this case is effectively ignored when inferring the device. I'd say the bug is actually fairly orthogonal to using generators.

https://github.com/deepinv/deepinv/blob/09be3a92aac0d08ef908c992a4515e5fea034e05/deepinv/physics/noise.py#L235-L240

Related: #536

**Comment by jscanvic** [comment]
There's a second bug: Denoising ignores the device silently. The root cause is that the device gets assigned to kwargs, which is passed down the inheritance chain and when it gets passed to LinearPhysics it is ignored instead of being passed down to Physics.

https://github.com/deepinv/deepinv/blob/09be3a92aac0d08ef908c992a4515e5fea034e05/deepinv/physics/forward.py#L395-L416

This is a major bug because every time an optional argument is passed to a LinearPhysics with the wrong name (possibly because we changed it in a recent update), the default value is going to be used silently instead of the user-provided value.

Besides that, we can choose between 1) adding the device argument in Physics which would basically amount to doing `self.to(device)` at the end of the constructor 2) not adding it and letting users rely on Physics.to themselves (kind of like the convention torch follows for layers & models).

I personally prefer option 2 but I don't mind option 1 either.

**Comment by jscanvic** [comment]
Related: #423

**Comment by mh-nguyen712** [comment]
The bug comes from the way we infer the device of the noise model: `device = getattr(rng, "device", "cpu")` 
So correct this would fix the bug.

**Comment by mh-nguyen712** [comment]
Note that we've already call `self.to(x.device)` in the `forward` of the noise model. 
So finally it will be always on the same device as input. This is a bit different than usual torch module. 
What do you think @jscanvic ? Related to your PR #637

**Comment by jscanvic** [comment]
I didn't notice that so yeah it's not an issue for inference still they should be on the right device right away I think

**Comment by mh-nguyen712** [comment]
I meant should we keep as it? Or should  we remove `self.to(device)` in forward and user have to call `noise_model.to(device)` or `physics.to(device)`, as we usually do in torch?

**Comment by jscanvic** [comment]
Ah sorry! I'm happy with keeping things as they are for now!
